### ISSUE TITLE
Headers not getting set in sync and replications in v3.3.1

### ISSUE THREAD

**Issue by alanquigley** [pouchdb/pouchdb#3543]
I'm setting the following options object

      var options = {
        ajax: {
          headers: {
             Authorisation : 'my token value'
          }
        }
      }; 

and passing it into the PouchDB constructor.

this.db = new PouchDB(_id, options);

But my sync and replication requests don't contain the header.

**Comment by nolanlawson** [comment]
I think you mean `Authorization` not `Authorisation`.

**Comment by alanquigley** [comment]
Uk spelling, but I've tried both, am I missing something?

**Comment by nolanlawson** [comment]
Looks like you may be missing the "Basic" keyword. Here's a live working example: http://bl.ocks.org/nolanlawson/61c9c615516582436e22

**Comment by alanquigley** [comment]
I'm trying to pass a jwt auth token through a proxy, the format is: 
Authorization:Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJ1c2VySUQiOiJvcmcuY291Y2hkYi51c2VyOmllX2FsYW5xdWlnbGV5X2dtYWlsX2NvbSJ9.X_qWaFXVdxMEhxsyfURALZrLbklTq03Gh33vA00Hbjg
Is it possible to set custom headers?

**Comment by alanquigley** [comment]
https://github.com/alanquigley/pouchdb/commit/c76c2a596ba97d2f58db7bd4ef3f153d503c8a23

The above tweak seems to fix my issue.

**Comment by alanquigley** [comment]
Got it working, the global settings don't seem to be working, If I put the headers in into the sync options it attaches them.

 
          var token = 'Bearer '+ AuthTokenFactory.getToken();
          var options = {
            ajax: {
              headers: {
                 Authorization : token
              }
            },
            filter : filter
          };

**Comment by nolanlawson** [comment]
Great, glad it's working!

**Comment by alanquigley** [comment]
The documentation seems to indicate that attaching the headers globally should work, is this a bug? or an error in the docs?

**Comment by nolanlawson** [comment]
You're right; that looks like a bug.

**Comment by alanquigley** [comment]
https://github.com/alanquigley/pouchdb/commit/c76c2a596ba97d2f58db7bd4ef3f153d503c8a23

When I said I got it working I was mistaken, I actually required the above code change to allow me to add  the options sync method.

**Comment by daleharvey** [comment]
So you can pass these options to the PouchDB constructor

`new PouchDB('hello', {ajax: { headers: { ....`

However not to the `.replicate` or `.sync` end points, the way to do this now is to 
```
var remote = new PouchDB('hello', opts);
location.sync(remote);
```

However enough people have expected this to work as you do that I am beginning to thing its worth just adding

**Comment by alanquigley** [comment]
The above makes sense, I just re-read the the guide section on replications, if I had followed it I don't think I would have made this mistake, but I followed the API usage example. Thanks

**Comment by nolanlawson** [comment]
We should document this in the API then. It makes sense to me that

```js
var db = new PouchDB('foo', {options...}
```

^ the options here don't apply to

```js
PouchDB.replicate
```

but they do apply to

```js
db.replicate
```

**Comment by mrded** [comment]
Attaching the headers globally doesn't work for me neither (pouchdb 3.3.1)

**Comment by nolanlawson** [comment]
I missed https://github.com/alanquigley/pouchdb/commit/c76c2a596ba97d2f58db7bd4ef3f153d503c8a23; the fix looks correct to me.

**Comment by mrded** [comment]
Problem not just with headers. All arguments to replication are ignored including views or filters.

    db.replicate.from(remoteDB, {filter: 'post/by_user'})

**Comment by nolanlawson** [comment]
Yup, you're right. But the fix should cover that.

**Comment by mrded** [comment]
Unfortunately, it's still not working for me.

**Comment by nolanlawson** [comment]
The fix is not merged yet. :)

**Comment by mrded** [comment]
I know, I build it myself, and it doesn't solve the problem.

**Comment by nolanlawson** [comment]
A reproducible test case would be very helpful. :)

**Comment by daleharvey** [comment]
> Problem is not just with headers. All arguments to replication are ignored including views or filters.

We dont currently support view replication or referencing design doc filters in the replicator

**Comment by mrded** [comment]
> We dont currently support view replication or referencing design doc filters in the replicator

However, replication by filters was working on 3.2.1 version.

**Comment by daleharvey** [comment]
Ah you are right, this got broken by https://github.com/pouchdb/pouchdb/commit/fe43de04557fd4df72ea9fe6b2e01c88e1ef8260

**Comment by mrded** [comment]
ok, what is latest version with working replication? :)

**Comment by nolanlawson** [comment]
@mrded If you're in a hurry, then [here's a build of pouchdb](https://nolanlawson.s3.amazonaws.com/pouchdb/www/20150220/pouchdb.min.js) using the latest code, but with fe43de0 reverted. We'll look into this.

**Comment by nolanlawson** [comment]
OK, I see what's going on. After fe43de0, only filter *function*s are accepted, i.e. all filtering is client-side. So you can pass in a JavaScript function as a filter and that will work, but passing in a `ddoc/view` name will not longer work.

**Comment by nolanlawson** [comment]
Think I found a fix. BTW to clarify, this is really two separate issues. One issue is that headers aren't passed in the `opts`, and the other issue is that `opts.filter` is ignored when it's a ddoc.

**Comment by mrded** [comment]
I confirm, that filtered replication with version without [fe43de0](https://github.com/pouchdb/pouchdb/commit/fe43de04557fd4df72ea9fe6b2e01c88e1ef8260) works as expected. Thanks @nolanlawson!

**Comment by nolanlawson** [comment]
fixed by 2e995c080c7b273fe3215d0e4cd289ac9b9a9203
